Bookworm installer by rodw

rodw-au opened the pull request:
The installer has been updated to follow the methods in the v2.9 getting started document I wrote.
ref: http://linuxcnc.org/docs/2.9/html/getting-started/getting-linuxcnc.html

This now uses the new buildbot at http://buildbot2.highlab.com/ so Linuxcnc 2.9 can be upgraded at any time

Please refer to notes in the read.ME and also in some of the changed files.

There are  a few things I would like to add as noted in the read.ME file but I don't fully understand apt so any help would be appreciated.

The main additions I would like from here is to add the repositories for: 
QTPYVCP:  https://kcjengr.github.io/probe_basic/quick_start_apt.html
Ethercat: https://forum.linuxcnc.org/ethercat/45336-ethercat-installation-from-repositories-how-to-step-by-step

Both of them add repos in a different way to Seb's new buildbot so I am a bit lost here. Help/PR's would be appreciated

A completed . ISO  as at 21 August 2023 (version 1092) can be downloaded from my Google Drive:
https://drive.google.com/file/d/16nprK_5HeURoNGd3tHWSVsxtHUGY3_S_/view?usp=drive_link

rodw-au committed:
First commit

rodw-au committed:
commit archives files

andypugh commented on auto/config:
```diff
@@ -21,7 +21,4 @@ lb config noauto \
-# It is currently Debian/testing, which do not have updates nor security
-# APT repo.
-#--parent-mirror-chroot-security http://deb.debian.org/debian-security
+# Debian 12 strips out non-free if it exists nad also removes non-free-firmware if installed to HDD
```
nad -> and

rodw-au replied:
Fixed in my repo

andypugh commented:
It would be useful to know if buildbot2 is there for the duration, or if the expectation is to retire some buildbot workers and move them there. @SebKuzminsky

Commit:
Update config

fix typo. Thanks @andy!

rodw-au commented:
@andypugh Changing the URL of the repository is a pretty trivial change and restricted to changes in the archives folder which are all in a separate commit in this PR. I always assumed the buildbot2 URL would becomes a Linuxcnc one eventually.

I will also have to do some updates to the 2.9 "getting linuxcnc" docs when this goes live.

rodw-au committed:
Added non-free repositories via prefs files because Debian Live
does not directly allow it

rodw-au committed:
Updated to include ethercat repositories.
For some reason a number of existing hooks are showing as new files
Just 500 and 510 have been aded

rodw-au committed:
Added QTPYVCP repositories.

rodw-au committed:
Removed preseed settings that interfered with auto/config file settings.
Removed seperate pref files for non-free becasue they have moveed to auto/config
Moved QTPYVCP repository to hook 0520 to avoid warnings when building

rodw-au committed:
Synced READEME.with git

rodw-au commented:
This is now finished as far as I can see.. Some of our pre-seed customizations broke the auto/config settings.

Third party repositories are configured as hooks 0510 and 0520 in config/hooks/normal so they are easy to remove if that becomes a requirement. We are not sharing anything that could be a licensing issue, we are just configuring some repositories as a convenience to our user base. Its up to them to meet any further requirements if they choose to use the repositories.

The only other third party repository (which does not exist yet) would be to the Linuxcnc-ethercat--driver. Ideally it would be hosted on a web site owned by a member of the Ethercat Technology Group to meet any licensing concerns.

It might be premature, but it would be good to commit these changes to the Bookworm branch sooner than later so there is an easy path to linuxcnc on Bookworm.

Commit:
Update README.md

Deleted todo's

rodw-au committed:
1. resolve warning from linuxcnc buildbot2 for non-existent i386 repos
2. Add repository for linuxcnc ethercat driver

petterreinholdtsen commented on auto/config:
```diff
@@ -6,22 +6,18 @@ lb config noauto \
 	--distribution bookworm \
 	--binary-images iso-hybrid \
-	--debian-installer netboot \
-	--archive-areas "main contrib non-free non-free-firmware" \
```
What is the point of this change?  It seem like a non-change to me, reordering non-free and non-free-firmware?

rodw-au replied:
I need to revisit this as there is some preseed script that overrides this and did not include non-free-firmware. I thought I deleted it. It may still be active based on recent tests. We need non-free because some network drivers such as the R8168-dkms are not firmware so remain in non-free.

rodw-au replied:
It appears I  had a regression here. and with the preseed

petterreinholdtsen commented on README.md:
```diff
@@ -6,6 +6,22 @@
     git checkout bookworm
     [ -x /usr/bin/lb ] || sudo apt install live-build
     lb config
-    sudo lb build
```
This is better done using sudo -i.

petterreinholdtsen commented on config/archives/linuxcnc-uspace.list.binary:
```diff
@@ -0,0 +1 @@
+deb [arch=amd64] http://buildbot2.highlab.com/debian/ bookworm 2.9-uspace
```
Why do you limit the arch here?

What about specifying the key to use like done with the repo added in config/hooks/normal/0520-Qtpyvcp-repos.hook.chroot? It would reduce the attack surface a bit.

rodw-au replied:
Currently, only amd64 is present on the new repo. When doing an apt update, without the arch flag, warnings are raised that i386 repos cannot be found. On review, this could be because of a malformed linuxcnc repository but I am out of my depth
 http://buildbot2.highlab.com/debian/dists/bookworm/2.9-uspace/ 

I did try to use the new apt sources format but got errors so reverted.

petterreinholdtsen commented on config/hooks/normal/0510-install-ethercat-sources.hook.chroot:
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh
+sudo mkdir -p /usr/local/share/keyrings/
+wget -O- https://build.opensuse.org/projects/science:EtherLab/signing_keys/download?kind=gpg | gpg --dearmor | sudo dd of=/usr/local/share/keyrings/science_EtherLab.gpg
+tee -a /etc/apt/sources.list.d/ighvh.sources > /dev/null <<EOT
```
Why are you using 'tee' with /dev/null here?  It is cleaner to use
```
cat >> /etc/apt/sources.list.d/ighvh.sources <<EOT
```

rodw-au replied:
Using the recommended script supplied by IGh

petterreinholdtsen commented on config/hooks/normal/0520-Qtpyvcp-repos.hook.chroot:
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
+curl sS https://repository.qtpyvcp.com/repo/kcjengr.key | gpg --dearmor | sudo tee /usr/share/keyrings/kcjengr.gpg
```
What is 'sS' doing here?  It causes curl to report 'Could not resolve host: sS'.

I believe the key should be stored in /etc/apt/keyrings/, not under /usr/, as it is not part of a installed package.

rodw-au replied:
The curl format is as provided by the qtpyvcp guys here: https://kcjengr.github.io/probe_basic/quick_start_apt.html
I did not get any errors here on several tests and the repos installed for me.

The standard gpg method relies on the users' .gnupg folder for keys. I did a lot of research to find out how to bypass the users profile because no user is present on install.
I standardised on /usr because it was indicated that data in /etc/apt/keyrings was blindly accepted by apt which could be a security risk. I am open to further guidance though.

rodw-au replied:
Ok, on further review, /etc/apt/keyrings/, is OK. I will revise locations.
Ref:  https://askubuntu.com/questions/1286545/what-commands-exactly-should-replace-the-deprecated-apt-key

rodw-au replied:
argh. Should have been -sS

petterreinholdtsen commented on config/hooks/normal/0520-Qtpyvcp-repos.hook.chroot:
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
+curl sS https://repository.qtpyvcp.com/repo/kcjengr.key | gpg --dearmor | sudo tee /usr/share/keyrings/kcjengr.gpg
+sudo gpg --homedir /tmp --no-default-keyring --keyring /usr/share/keyrings/QTPYVCP.gpg --keyserver keys.openpgp.org --recv-keys 2DEC041F290DF85A
+sudo echo "deb [arch=amd64 signed-by=/usr/share/keyrings/QTPYVCP.gpg] https://repository.qtpyvcp.com/apt develop main " | sudo tee -a /etc/apt/sources.list.d/qtpyvcp.list >/dev/null
```
The 'sudo' in front of the echo command serve no purpose.

rodw-au replied:
Again I used the QTPYVCP supplied format. But yes we're are root so can be cleaned up.

rodw-au replied:
Moved to /etc/apt/keyrings as suggested

petterreinholdtsen commented on config/hooks/normal/0530-linuxcnc-ethercat-sources.hook.chroot:
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
+curl -s https://packagecloud.io/install/repositories/rodw-au/rodw-au/script.deb.sh | sudo bash
```
Downloading random stuff from the Internet and running it as root seem like a very bad idea to me.  This could compromise the build host and any images generated at the whim of those controlling the result from the URL provided.

rodw-au replied:
I can change this to be a more normal format. It is a very cool script though!

rodw-au replied:
This has now been revised to a local script (hook 530)

petterreinholdtsen commented on config/hooks/normal/1000-create-mtab-symlink.hook.chroot:
```diff
@@ -0,0 +1 @@
+/usr/share/live/build/hooks/normal/1000-create-mtab-symlink.hook.chroot
```
Why are all these symlinks needed?  Are they not added by lb config?

rodw-au replied:
That is the original code, Maybe I bumped it by accident.
Our scripts are in the 0500-0599 range.

petterreinholdtsen reviewed (changes requested):
Believe the change need some adjustments before it can go it.  I strongly recommend splitting the changes into separate pull requests, one for each feature (say for new packages, enabling individual extra repositories, enabling security, enabling updates), to allow each feature to be evaluated individually without delaying the inclusion of other fixes.

rodw-au commented on config/includes.installer/preseed.cfg:
I don't know why lines 7 & 8 we're added back in. Must have been a glitch. They should be deleted so the config file settings are respected.

rodw-au replied:
modified this several times chasing my tail to get the mirrors and sources.list set up correctly. Now final!

rodw-au committed:
Update preseed and auto/config to put non-free and non-free-firmware in sources.list

rodw-au committed:
removed stash

rodw-au committed:
Minor edits and 3 possible run time dependencies

rodw-au committed:
stop installing brltty and speech-dispatcher

rodw-au committed:
Remove packagecloud.io online script, Replace with local script

rodw-au committed:
Modify preseed file to get proxy and sources.list initiated correctly

rodw-au committed:
Modified typo in hook #520

rodw-au committed:
Moved key to /etc/apt/keyrings as requested

rodw-au commented:
This is now complete for use for the Linuxcnc 2.9 release.
Test lightly with Axis, Gmocappy and QTdragon_HD, all open without errors ands appear to run.

The only outstanding issue I would have liked to do is to deploy an instructions document to the user's home directory.
I was led to believe that files in config/includes.installer/etc/skel would be deployed to the user's home directory. If somebody could review and resolve this, it would be fantastic.

rodw-au committed:
Revert 0520 hook. Could not get qtpyvcp to work in /etc/apt folders

rodw-au committed:
save qtpyvcp keys in the /etc/apt/keyrings folder

rodw-au committed:
Revised 510 hook to use /etc/apt folders

rodw-au committed:
Add some extra packages for key handling

rodw-au committed:
Use /etc/apt folders

rodw-au committed:
Final changes to use /etc/apt folders

rodw-au committed:
move 0510 hook (Ethercat repos) to /etc/apt/trusted.gpg.d

rodw-au committed:
Modified Hook 0520 (QTPYVCP) to use /etc/apt/trusted/gpg.d

rodw-au commented:
This is now complete for use for the Linuxcnc 2.9 release.
Test lightly with Axis, Gmocappy and QTdragon_HD, all open without errors ands appear to run.

rodw-au committed:
Added README.md to all Users Home folder with basic instructions

smoe commented:
This is all some very impressive nitty-gritty work you have accomplished. Many thanks!

rodw-au commented:
Thanks Steffen, there was a lot to learn to get it to this final stage.

andypugh commented:
Do we really want to be pulling all the debs from the buildbot rather than the LinuxCNC repository? 
@SebKuzminsky Is this a sensible thing to do? 
(Note that as I write this, I am busy transferring a set of debs to the LinuxCNC repository)

rodw-au commented:
When you have updated the repository, please let me know details and I will update the installer BUT:

Given that v 2.9 will essentially be locked following  its release and it will be pushed to Debian,
I think  a better approach would be to update the APT pinning to pull our debs from Debian unstable branch. That would give best performance as Debian will have better performing infrastructure then us.

If we do eventually get into Debian 12.3, the pinning could be removed altogether.

smoe commented:
The Debian testing branch may be what we are after - that is unstable plus 5 days. When we build and test against unstable, this would allow us to spot incompatible libraries very early when packages are still in unstable and we would then issue a bug report that would prevent that incompatible library from transitioning to testing.

rodw-au commented:
Thanks Steffan,
I built it against unstable today but I got called away before I committed it.
I'll redo it in another branch against testing tomorrow and you can choose which one to use.
I might not be able to test the installers until tomorrow night.
